File: lib/go/cache.go
package radixip

import (
	"context"
	"encoding/json"
	"net"
	"sync"
	"time"
)

// CacheConfig holds configuration for the cache
type CacheConfig struct {
	MaxEntries int
	TTLSeconds *uint64
}

// RadixCache manages caching between Redis and the engine
type RadixCache struct {
	cache   sync.RWMutex
	entries map[string]*cacheEntry // key is IP string
	config  CacheConfig
	engine  RadixEngine
	redis   *RedisClient
}

type cacheEntry struct {
	metadata  *Metadata
	expiresAt *time.Time
}

// NewRadixCache creates a new RadixCache instance
func NewRadixCache(config CacheConfig, engine RadixEngine, redisClient *RedisClient) *RadixCache {
	rc := &RadixCache{
		entries: make(map[string]*cacheEntry),
		config:  config,
		engine:  engine,
		redis:   redisClient,
	}

	// Boot-load prefixes from Redis
	if rc.redis != nil {
		ctx := context.Background()
		entries, err := rc.redis.HGetAll(ctx, "radixip:entries")
		if err == nil {
			for cidr, metaJSON := range entries {
				ipNet, _, err := net.ParseCIDR(cidr)
				if err == nil {
					network := IpNetwork{IP: ipNet, Mask: net.CIDRMask(24, 32)} // Simplify mask logic
					if _, ipnet, err := net.ParseCIDR(cidr); err == nil {
						network = IpNetwork{IP: ipnet.IP, Mask: ipnet.Mask}
					}
					var meta Metadata
					if json.Unmarshal([]byte(metaJSON), &meta) == nil {
						// Convert to *net.IPNet
						ipNetObj := &net.IPNet{
							IP:   network.IP,
							Mask: network.Mask,
						}
						rc.engine.Insert(ipNetObj, meta)
					}
				}
			}
		}
	}

	return rc
}

// lookupWithCache performs a lookup with caching
func (c *RadixCache) lookupWithCache(ip net.IP) *Metadata {
	ipStr := ip.String()

	// Check cache first (read lock)
	c.cache.RLock()
	if entry, exists := c.entries[ipStr]; exists {
		// Check TTL if configured
		if c.config.TTLSeconds != nil && entry.expiresAt != nil {
			if time.Now().Before(*entry.expiresAt) {
				c.cache.RUnlock()
				return entry.metadata
			}
			// Expired - we'll remove it with write lock
			c.cache.RUnlock()
			c.cache.Lock()
			delete(c.entries, ipStr)
			c.cache.Unlock()
		} else {
			// No TTL, return cached value
			c.cache.RUnlock()
			return entry.metadata
		}
	} else {
		c.cache.RUnlock()
	}

	// Cache miss - query engine
	result := c.engine.Lookup(ip)

	// If not found in engine, check Redis lookup cache
	if result == nil && c.redis != nil {
		ctx := context.Background()
		cachedMeta, err := c.redis.Get(ctx, "radixip:lookup:"+ipStr)
		if err == nil && cachedMeta != "" {
			var meta Metadata
			if json.Unmarshal([]byte(cachedMeta), &meta) == nil {
				result = &meta
			}
		}
	}

	// Store in cache
	c.cache.Lock()
	defer c.cache.Unlock()

	// Check if we need to evict
	if len(c.entries) >= c.config.MaxEntries {
		// Simple eviction - remove oldest (first key)
		for key := range c.entries {
			delete(c.entries, key)
			break
		}
	}

	// Store the result (even if nil)
	var expiresAt *time.Time
	if c.config.TTLSeconds != nil {
		t := time.Now().Add(time.Duration(*c.config.TTLSeconds) * time.Second)
		expiresAt = &t
	}

	c.entries[ipStr] = &cacheEntry{
		metadata:  result,
		expiresAt: expiresAt,
	}

	// Also store in Redis lookup cache
	if c.redis != nil && result != nil {
		ctx := context.Background()
		if jsonData, err := json.Marshal(result); err == nil {
			c.redis.Set(ctx, "radixip:lookup:"+ipStr, string(jsonData))
		}
	}

	return result
}

// invalidate removes cache entries that match the given network prefix
func (c *RadixCache) invalidate(prefix *net.IPNet) {
	c.cache.Lock()
	defer c.cache.Unlock()

	for ipStr := range c.entries {
		ip := net.ParseIP(ipStr)
		if ip != nil && networkContainsIP(prefix, ip) {
			delete(c.entries, ipStr)
		}
	}
}

// clear removes all cache entries
func (c *RadixCache) clear() {
	c.cache.Lock()
	defer c.cache.Unlock()
	c.entries = make(map[string]*cacheEntry)
}

// CachedEngine wraps a RadixEngine with caching functionality
type CachedEngine struct {
	inner RadixEngine
	cache *RadixCache
}

// NewCachedEngine creates a new CachedEngine instance
func NewCachedEngine(inner RadixEngine, config CacheConfig, redisClient *RedisClient) *CachedEngine {
	cache := NewRadixCache(config, inner, redisClient)
	return &CachedEngine{
		inner: inner,
		cache: cache,
	}
}

// Insert adds a prefix with metadata and invalidates cache
func (e *CachedEngine) Insert(prefix *IpNetwork, metadata Metadata) error {
	// Convert to *net.IPNet
	ipNetObj := &net.IPNet{
		IP:   prefix.IP,
		Mask: prefix.Mask,
	}
	if err := e.inner.Insert(ipNetObj, metadata); err != nil {
		return err
	}

	// Persist to Redis
	if e.cache.redis != nil {
		ctx := context.Background()
		if jsonData, err := json.Marshal(metadata); err == nil {
			e.cache.redis.HSet(ctx, "radixip:entries", prefix.String(), string(jsonData))
		}
	}

	// Invalidate relevant cache entries
	netPrefix := net.IPNet{IP: prefix.IP, Mask: prefix.Mask}
	e.cache.invalidate(&netPrefix)
	return nil
}

// Lookup performs a lookup with caching
func (e *CachedEngine) Lookup(ip net.IP) *Metadata {
	return e.cache.lookupWithCache(ip)
}

// Remove deletes a prefix and invalidates cache
func (e *CachedEngine) Remove(prefix *IpNetwork) *Metadata {
	// Convert to *net.IPNet
	ipNetObj := &net.IPNet{
		IP:   prefix.IP,
		Mask: prefix.Mask,
	}
	result := e.inner.Remove(ipNetObj)

	// Remove from Redis
	if e.cache.redis != nil {
		ctx := context.Background()
		e.cache.redis.HDel(ctx, "radixip:entries", prefix.String())
	}

	netPrefix := net.IPNet{IP: prefix.IP, Mask: prefix.Mask}
	e.cache.invalidate(&netPrefix)
	return result
}

// Contains checks if a prefix exists
func (e *CachedEngine) Contains(prefix *IpNetwork) bool {
	// Convert to *net.IPNet
	ipNetObj := &net.IPNet{
		IP:   prefix.IP,
		Mask: prefix.Mask,
	}
	return e.inner.Contains(ipNetObj)
}

// Clear removes all entries and clears cache
func (e *CachedEngine) Clear() {
	e.inner.Clear()
	e.cache.clear()
}

// Size returns the number of entries in the underlying engine
func (e *CachedEngine) Size() int {
	return e.inner.Size()
}

// Stats returns engine statistics
func (e *CachedEngine) Stats() *EngineStats {
	return e.inner.Stats()
}

// Helper function to check if an IP is within a network prefix
func networkContainsIP(prefix *net.IPNet, ip net.IP) bool {
	if prefix.IP.To4() != nil && ip.To4() != nil {
		// IPv4
		return prefix.Contains(ip)
	}
	if prefix.IP.To16() != nil && ip.To16() != nil {
		// IPv6
		return prefix.Contains(ip)
	}
	return false
}

File: lib/rust/src/cache.rs
use super::traits::*;
use crate::lpm::network_contains_ip;
use crate::types::{EngineStats, Metadata};
use ipnetwork::IpNetwork;
use std::collections::HashMap;
use std::net::IpAddr;
use std::sync::{Arc, RwLock};

// Cache configuration between redis and engine
pub struct CacheConfig {
    pub max_entries: usize,
    pub ttl_seconds: Option<u64>,
}

pub struct RadixCache {
    cache: RwLock<HashMap<IpAddr, Option<Metadata>>>,
    config: CacheConfig,
    engine: Arc<dyn RadixEngine>,
    #[cfg(feature = "redis")]
    redis: Option<crate::redis::RedisClient>,
}

impl RadixCache {
    #[cfg(not(feature = "redis"))]
    pub fn new(config: CacheConfig, engine: Arc<dyn RadixEngine>) -> Self {
        Self {
            cache: RwLock::new(HashMap::new()),
            config,
            engine,
        }
    }

    #[cfg(feature = "redis")]
    pub fn new(
        config: CacheConfig,
        engine: Arc<dyn RadixEngine>,
        redis: Option<crate::redis::RedisClient>,
    ) -> Self {
        let cache = Self {
            cache: RwLock::new(HashMap::new()),
            config,
            engine: engine.clone(),
            redis,
        };

        // Boot-load prefixes from Redis into engine
        if let Some(r) = &cache.redis {
            if let Ok(entries) = r.hgetall_sync("radixip:entries") {
                for (cidr, meta_json) in entries {
                    if let Ok(network) = cidr.parse::<IpNetwork>() {
                        if let Ok(metadata) = serde_json::from_str::<Metadata>(&meta_json) {
                            let _ = engine.insert(network, metadata);
                        }
                    }
                }
            }
        }
        cache
    }

    pub fn lookup_with_cache(&self, ip: &IpAddr) -> Option<Metadata> {
        // Check cache first
        if let Some(entry) = self.cache.read().unwrap().get(ip) {
            return entry.clone();
        }

        // Cache miss - query engine
        let mut result = self.engine.lookup(ip);

        // If not found in engine, and redis is enabled, we could query Redis.
        // Wait, the engine has been boot-loaded with all prefixes!
        // So if it's not in the engine, it's not in Redis either.
        // However, if we want to query a shared lookup cache in Redis:
        #[cfg(feature = "redis")]
        if result.is_none() {
            if let Some(r) = &self.redis {
                let key = format!("radixip:lookup:{}", ip);
                if let Ok(Some(cached_meta)) = r.get_sync(&key) {
                    if let Ok(meta) = serde_json::from_str::<Metadata>(&cached_meta) {
                        result = Some(meta);
                    }
                }
            }
        }

        // Store in cache
        let mut cache = self.cache.write().unwrap();
        if cache.len() >= self.config.max_entries {
            // Simple eviction - remove oldest
            if let Some(key) = cache.keys().next().cloned() {
                cache.remove(&key);
            }
        }
        cache.insert(ip.clone(), result.clone());

        // Also store in Redis lookup cache
        #[cfg(feature = "redis")]
        if let (Some(r), Some(res)) = (&self.redis, &result) {
            if let Ok(json) = serde_json::to_string(res) {
                let key = format!("radixip:lookup:{}", ip);
                let _ = r.set_sync(&key, &json);
            }
        }

        result
    }

    pub fn invalidate(&self, prefix: &IpNetwork) {
        // Remove entries that match prefix
        let mut cache = self.cache.write().unwrap();
        cache.retain(|ip, _| !network_contains_ip(prefix, ip));
    }

    pub fn clear(&self) {
        self.cache.write().unwrap().clear();
    }
}

// Cache wrapper that implements RadixEngine
pub struct CachedEngine {
    inner: Arc<dyn RadixEngine>,
    cache: RadixCache,
}

impl CachedEngine {
    #[cfg(not(feature = "redis"))]
    pub fn new(inner: Arc<dyn RadixEngine>, config: CacheConfig) -> Self {
        let cache = RadixCache::new(config, inner.clone());
        Self { inner, cache }
    }

    #[cfg(feature = "redis")]
    pub fn new(
        inner: Arc<dyn RadixEngine>,
        config: CacheConfig,
        redis: Option<crate::redis::RedisClient>,
    ) -> Self {
        let cache = RadixCache::new(config, inner.clone(), redis);
        Self { inner, cache }
    }
}

impl RadixEngine for CachedEngine {
    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        self.inner.insert(prefix, metadata.clone())?;

        // Persist to Redis
        #[cfg(feature = "redis")]
        if let Some(r) = &self.cache.redis {
            if let Ok(json) = serde_json::to_string(&metadata) {
                let _ = r.hset_sync("radixip:entries", &prefix.to_string(), &json);
            }
        }

        // Invalidate relevant cache entries
        self.cache.invalidate(&prefix);
        Ok(())
    }

    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        self.cache.lookup_with_cache(ip)
    }

    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        let result = self.inner.remove(prefix);

        // Remove from Redis
        #[cfg(feature = "redis")]
        if let Some(r) = &self.cache.redis {
            let _ = r.hdel_sync("radixip:entries", &prefix.to_string());
        }

        self.cache.invalidate(prefix);
        result
    }

    fn contains(&self, prefix: &IpNetwork) -> bool {
        self.inner.contains(prefix)
    }

    fn clear(&self) {
        self.inner.clear();
        self.cache.clear();
    }

    fn size(&self) -> usize {
        self.inner.size()
    }

    fn stats(&self) -> EngineStats {
        self.inner.stats()
    }
}
